### ISSUE TITLE
number of neighbours returned far too lower than K

### ISSUE THREAD

**Issue by abmitra84** [nmslib/nmslib#490]
Hi @searchivarius ,

I have a dataset with 1 mn * 2000 dimension - all numeric fields. At the index time, I am using following params
```
M=20
efC=3000
num_threads=12
```

and at the time of search following params are used
```
K=30000
efS=3000
num_threads=12
```

```
method=hnsw,
space=angulardist
```

I am consistently getting about 3000+ neighbours. If I understood the `hnsw` clearly `efC` or `efS` shouldn't control how many neighbours I will get as returned. Or am I wrong ?   

Please let me know what I need to change to ensure to get 30000 neighbours

**Comment by searchivarius** [comment]
Hi @abmitra84 this has been discussed before and gitter is a better place to discuss such issues:https://gitter.im/nmslib/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge. 

There's unfortunately no guarantee (currently) that we produce that many number. This is especially true for large K. You can try to increase efC to about 1000 and efS to say 5000. They do control (indirectly) the number of nearest neighbors.

Also, for the very large # of neighbors (and a small dataset), it may be better to run a bruteforce search or use a different approach (NAPP).

**Comment by searchivarius** [comment]
I am closing it, b/c it's a duplicate issue.

**Comment by abmitra84** [comment]
Thanks @searchivarius . I see this has been discussed earlier, but going through the paper the relation (implicit or otherwise) and the nature of relation among efC/efS vs K was not entirely clear to me. Thanks for your input, appreciate quick response.

**Comment by searchivarius** [comment]
You are correct: This relationship is not well explored. Off the top of the head, efS defines the depth of queue. If the depth is not large enough you may stop finding k nearest neighbors too early. As a second thought, I think one may try to modify the stop condition to prevent this from happening in a brute-force fashion. This might affect the speed somewhat. However, now we have a proper testing suit, we can control for this regression. What do you think?

**Comment by abmitra84** [comment]
@searchivarius , yes we can simply put a condition of `max(K, len(queue))` for stop condition. In my set up an `efS=3000` would consistently generate ~3500 neighbours. I had to set `efS= 1.1 * K` to get consistent K=30000 neighbours. I also modified 'M=50' and 'efC=100*M' but I didn't feel this set up really impacts the number of neighbours returned.

On another note, in your previous comment you mentioned NAPP, also you mentioned smaller dataset, did you mean NAPP performs well with smaller dataset for larger neighbours ?

**Comment by searchivarius** [comment]
NAPP's performance is overall quite reasonable, it would be interesting to compare against your settings for efS and efC, as well as with the fixed stopping condition, will you give it a shot? We would have three things to compare:
1. Outcomes where you bump up M and efC
2. Outcomes with NAPP
3. Outcomes with a fixed stop condition (for both indexing and searching algos).

**Comment by abmitra84** [comment]
1. and 2. I will do and let you know on this thread. Not sure how to implement the 3. I am using python and I am not good with C++ to make fundamental changes to the core part of the library. Can I do 3 in a different way just by using python API ?

**Comment by abmitra84** [comment]
@searchivarius , do you have advice for me to choose optimum range of parameter values for NAPP. 

I read the manual but still not sure about the right set of parameters against sample_size, feature dims and K. 

Also is there a set of papers I can read to understand how does it work?

**Comment by searchivarius** [comment]
@abmitra84 
3. I would need to patch the library. Would you test a dev branch?
4. NAPP parameters. 
  i.  Number of pivots `numPivot` should be relatively large, say 2000 
  ii. Number of indexed pivots is roughly a square root of this: `numPivotIndex` should be in the range 50 to 100.
  iii. Then you can play with various values of `numPivotSearch` at query time. Something in the range 3 to 10 should be reasonable.

**Comment by abmitra84** [comment]
Unfortunately I won't be able to test a dev branch with the data I have  (compliance restrictions). I can test with the datasets used by Ann Benchmarks

**Comment by searchivarius** [comment]
@abmitra84 not even locally on your servers?

**Comment by abmitra84** [comment]
@searchivarius I can't clone the dev branch to my workspace for legal restrictions.

**Comment by searchivarius** [comment]
@abmitra84 what about the master branch? Or you can only pip install?

**Comment by abmitra84** [comment]
@searchivarius  I can only do `pip install` when it is published through pypi or anaconda.
